File: src/frontend/src/services/MetadataService.ts
import { GetMetadataProfilesReply } from "@common/api/metadata/MetadataCommands";
import { MetadataProfileContainer } from "@common/data/entities/metadata/MetadataProfileContainer";
import { Service } from "@common/services/Service";
import { ColorTable } from "@common/ui/components/propertyeditor/utils/ColorTable";
import { deepClone, shortenDataStrings } from "@common/utils/ObjectUtils";

import { FrontendComponent } from "@/component/FrontendComponent";
import { FrontendServiceContext } from "@/services/FrontendServiceContext";

/**
 * Creates the metadata service.
 *
 * @param comp - The main component instance.
 *
 * @returns - The newly created service.
 */
export default function (comp: FrontendComponent): Service {
    function printableProfile(profile: MetadataProfileContainer): string {
        let obj = deepClone<MetadataProfileContainer>(profile);
        return JSON.stringify(shortenDataStrings(obj));
    }

    return comp.createService(
        "Metadata service",
        (svc: Service) => {
            svc.messageHandler(GetMetadataProfilesReply, (msg: GetMetadataProfilesReply, ctx: FrontendServiceContext) => {
                if (msg.success) {
                    ctx.logger.debug("Retrieved metadata profiles", "metadata", { profiles: msg.profiles.map(printableProfile) });

                    // @ts-ignore
                    ctx.metadataStore.profiles = msg.profiles;

                    ColorTable.populateFromProfileContainerList(ctx.metadataStore.profiles);
                } else {
                    ctx.logger.error("Unable to retrieve the metadata profiles", "metadata", { reason: msg.message });
                }
            });
        },
        FrontendServiceContext
    );
}

File: src/server/services/metadata_service.py
from common.py.api import GetMetadataProfilesCommand, GetMetadataProfilesReply
from common.py.services import Service

from .server_service_context import ServerServiceContext
from ..component import ServerComponent


def create_metadata_service(comp: ServerComponent) -> Service:
    """
    Creates the session service that handles all metadata/profiles messages.

    Args:
        comp: The main component instance.

    Returns:
        The newly created service.
    """

    svc = comp.create_service("Metadata service", context_type=ServerServiceContext)

    @svc.message_handler(GetMetadataProfilesCommand)
    def server_timeout(
        msg: GetMetadataProfilesCommand, ctx: ServerServiceContext
    ) -> None:
        GetMetadataProfilesReply.build(
            ctx.message_builder, msg, profiles=comp.server_data.profile_containers
        ).emit()

    return svc
